File: src/csharp/oop/MovieTheater/Cinemas/MovieSystem.cs
using MovieTheater.Cinemas.Screenings;

namespace MovieTheater.Cinemas;

public class MovieSystem
{
    private readonly List<Cinema> _cinemas = new();
    private readonly List<Movie> _movies = new();
    private readonly Dictionary<Movie, HashSet<Cinema>> _theatersByMovie = new();
    
    public IReadOnlyCollection<Cinema> GetTheaters() => _cinemas;
    
    public IReadOnlyCollection<Movie> GetMovies() => _movies;
    
    public IReadOnlyCollection<Cinema> GeTheatersByMovie(Movie movie)
        => _theatersByMovie.TryGetValue(movie, out var theaters) ? theaters : [];
    
    public void AddCinema(Cinema cinema)
    {
        if (_cinemas.Contains(cinema))
        {
            return;
        }
        
        _cinemas.Add(cinema);
        
        cinema.ScreeningChanged += TheaterOnScreeningChanged;
        
        foreach (var movie in cinema.Movies)
        {
            if (!_theatersByMovie.TryGetValue(movie, out var theaters))
            {
                _theatersByMovie[movie] = theaters = [];    
            }

            theaters.Add(cinema);
        }
    }

    public void RemoveCinema(Cinema cinema)
    {
        cinema.ScreeningChanged -= TheaterOnScreeningChanged;

        foreach (var movie in cinema.Movies)
        {
            if (_theatersByMovie.TryGetValue(movie, out var theaters))
            {
                theaters.Remove(cinema);    
            }
        }
        
        _cinemas.Remove(cinema);
    }

    public void AddMovie(Movie movie)
    {
        if (!_movies.Contains(movie))
        {
            _movies.Add(movie);
        }
    }

    public void RemoveMovie(Movie movie) => _movies.Remove(movie);
    
    private void TheaterOnScreeningChanged(object? sender, (Screening screening, bool added) args)
    {
        if (!_theatersByMovie.TryGetValue(args.screening.Movie, out var theaters) || sender is not Cinema theater)
        {
            return;
        }

        if (args.added)
        {
            theaters.Add(theater);
        }
        else
        {
            theaters.Remove(theater);
        }
    }
}

File: src/python/oop/MovieTheater/theater/movie_system.py

from .cinema import Cinema
from .movie import Movie


class MovieSystem:
    def __init__(self, cinemas: list[Cinema]):
        self._cinema = cinemas

        self._cinema_by_name = {}
        self._cinema_by_movie = {}
        self._movies = {}
        for cinema in cinemas:
            self._cinema_by_name[cinema.name] = cinema
            for movie in cinema.get_movies():
                if movie.title not in self._movies:
                    self._movies[movie.title] = movie

                if movie.title not in self._cinema_by_movie:
                    self._cinema_by_movie[movie.title] = []

                self._cinema_by_movie[movie.title].append(cinema)

    def get_cinema(self, name: str) -> Cinema:
        return self._cinema_by_name[name]

    def get_movies(self) -> list[Movie]:
        return list(self._movies.values())

    def get_cinemas_by_movie(self, movie: Movie) -> list[Cinema]:
        return list(self._cinema_by_movie[movie.title])
